sdk/core/System.TypeSpec/src/PipelineOptions.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ComponentModel;
+using System.Diagnostics;
+using System.ServiceModel.Rest.Core;
+using System.Threading;
+
+namespace System.ServiceModel.Rest;
+
+// Make options freezable
+public class PipelineOptions
+{
+    private object _lock = new object();
+    private MessagePipeline? _pipeline;
+    private volatile bool _isFrozen;
+
+    private IPipelinePolicy<PipelineMessage>? _retryPolicy;
+    private IPipelinePolicy<PipelineMessage>? _loggingPolicy;
+    private PipelineTransport<PipelineMessage> _transport;
+    private ReadOnlyMemory<IPipelinePolicy<PipelineMessage>> _perTryPolicies;
+    private ReadOnlyMemory<IPipelinePolicy<PipelineMessage>> _perCallPolicies;
+
+    public PipelineOptions(PipelineTransport<PipelineMessage> transport)
+    {
+        _transport = transport;
+    }
+
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public virtual MessagePipeline GetPipeline()
+    {
+        if (_pipeline != null) return _pipeline;
+
+        lock (_lock) {
+            if (IsFrozen) {
+                Debug.Assert(_pipeline != null);
+                return _pipeline!;
+            }
+            _pipeline = MessagePipeline.Create(this);
+            Freeze();
+            return _pipeline;
+        }
+    }
+
+    protected bool IsFrozen => _isFrozen;
+    protected void Freeze() => _isFrozen = true;
+
+    public ReadOnlyMemory<IPipelinePolicy<PipelineMessage>> PerTryPolicies {
```

Comment:
Making this ROM<policy> means a caller wanting to add/remove policies has to reallocate the array of polices to add to them.  Are we willing to make this usability tradeoff?

Reply:
Users should not be tweaking these options. They should set properties once and then never touch this type.

Reply:
If that's the case, how are we thinking callers would add policies to a pipeline for the duration of a single-method invocation?  e.g. to make a pipeline we would cache in RequestOptions?

I'm looking at this now, and it's a bit challenging to get right with the current API, e.g.:

```csharp
      // We need to pass the pipeline options to the client constructor
      // so that the auth policy gets added, otherwise we won't have it later.
      PipelineOptions pipelineOptions = new PipelineOptions();

      MapsClient client = new MapsClient(new Uri("https://atlas.microsoft.com"), credential, pipelineOptions);

      // Add a custom policy to the pipeline just for the one method
      pipelineOptions.PerCallPolicies = new PipelinePolicy[1];
      pipelineOptions.PerCallPolicies[0] = new CustomPolicy();
      RequestOptions options = new RequestOptions(pipelineOptions);

      IPAddress ipAddress = IPAddress.Parse("2001:4898:80e8:b::189");
      Result result = client.GetCountryCode(ipAddress.ToString(), options);
```

In the above example, setting the custom policy will overwrite the list with the auth policy in it.
Having an Add method would let us side-step that issue.

But, in [Azure.Core's RequestContext](https://apiview.dev/Assemblies/Review/77bdbcef49a3465fa96fac3fb745bf57#Azure.RequestContext), we have an AddPolicy method.  Does that make sense here?  It would let us freeze a pipeline in PipelineOptions and be clear when we're changing things one-off for a method and need to cache it in the RequestOptions instead

Reply:
Added sample-in-progress of this here: https://github.com/Azure/azure-sdk-for-net/pull/39740/files#diff-f9a63f2db2518d5664bb985363bc3959eff03bb93eaa5584e983beeea9b2b8b5R46-R48.

This should fail with a 401 because the auth policy is missing.
Actually, in this version, it fails because it's trying to modify a pipeline that's frozen, but once I make it "change-tracking," it should fail with the 401 instead.

Reply:
Addressed.